### PR TITLE
fix!: refactor structs to always be property eqs

### LINKED ISSUE #3015: wrong transformation from bigquery struct to presto row
**Before you file an issue**
- Make sure you specify the "read" dialect eg. parse_one(sql, read="spark")
- Check if the issue still exists on main 

**Fully reproducible code snippet**
```
>>> sqlglot.transpile("""select struct(true as sign, 1 as value)
... """, read="bigquery", write="presto")[0]
'SELECT ROW(TRUE AS sign, 1 AS value)'
```
and
```
trino> SELECT ROW(TRUE AS sign, 1 AS value)
    -> ;
Query 20240129_170954_00005_56ath failed: line 1:17: mismatched input 'AS'. Expecting: '%', ')', '*', '+', ',', '-', '.', '/', 'AND', 'AT', 'OR', 'ORDER', '[', '||', <predicate>
SELECT ROW(TRUE AS sign, 1 AS value)
```

**Official Documentation**
Please include links to official SQL documentation related to your issue.

### PR DESCRIPTION
**Pull request by tobymao**
closes #3015, allowing presto to have named structs if types are known

### CODE REVIEW

File: tests/dialects/test_bigquery.py [hunk 1]
```diff
@@ -18,6 +18,17 @@ class TestBigQuery(Validator):
...
+            "SELECT STRUCT(1, 2, 3), STRUCT(), STRUCT('abc'), STRUCT(1, t.str_col), STRUCT(1 as a, 'abc' AS b), STRUCT(str_col AS abc)",
+            write={
+                "bigquery": "SELECT STRUCT(1, 2, 3), STRUCT(), STRUCT('abc'), STRUCT(1, t.str_col), STRUCT(1 AS a, 'abc' AS b), STRUCT(str_col AS abc)",
+                "duckdb": "SELECT {'_0': 1, '_1': 2, '_2': 3}, {}, {'_0': 'abc'}, {'_0': 1, '_1': t.str_col}, {'a': 1, 'b': 'abc'}, {'abc': str_col}",
+                "spark": "SELECT STRUCT(1, 2, 3), STRUCT(), STRUCT('abc'), STRUCT(1, t.str_col), STRUCT(1 AS a, 'abc' AS b), STRUCT(str_col AS abc)",
+                "snowflake": "SELECT OBJECT_CONSTRUCT('', 1, '', 2, '', 3), OBJECT_CONSTRUCT(), OBJECT_CONSTRUCT('', abc), OBJECT_CONSTRUCT('', 1, '', str_col), OBJECT_CONSTRUCT('a', 1, 'b', 'abc'), OBJECT_CONSTRUCT('abc', str_col)",
+                # struct is a fallback without type inference
+                "trino": "SELECT ROW(1, 2, 3), ROW(), ROW('abc'), ROW(1, t.str_col), CAST(ROW(1, 'abc') AS ROW(a INTEGER, b VARCHAR)), STRUCT(str_col AS abc)",
+            },
+        )
```
Comment (georgesittas) [inline tests/dialects/test_bigquery.py:31]: Note that:

- `STRUCT()` is generally not supported I think in these dialects, only Snowflake allowed `OBJECT_CONSTRUCT()`
- This is invalid Snowflake `OBJECT_CONSTRUCT('', 1, '', 2, '', 3)` due to duplicate keys

Not sure if we wanna do something about this, just observing.

File: tests/dialects/test_bigquery.py [hunk 2]
```diff
@@ -603,11 +614,11 @@ def test_bigquery(self):
             read={"spark": "select posexplode_outer([])"},
         )
         self.validate_all(
-            "SELECT AS STRUCT ARRAY(SELECT AS STRUCT b FROM x) AS y FROM z",
+            "SELECT AS STRUCT ARRAY(SELECT AS STRUCT 1 AS b FROM x) AS y FROM z",
```
Comment (georgesittas) [inline tests/dialects/test_bigquery.py:617]: Did this need to change? I think both this and the previous one are valid BQ queries, but I'm curious if the old one can't be transpiled for some reason?

File: tests/dialects/test_duckdb.py [hunk 3]
```diff
@@ -55,8 +55,15 @@ def test_duckdb(self):
...
-            self.validate_all(struct_value, write={"presto": UnsupportedError})
+        self.validate_all(
+            "{'a': 1, 'b': '2'}", write={"presto": "CAST(ROW(1, '2') AS ROW(a INTEGER, b VARCHAR))"}
+        )
+        self.validate_all(
+            "struct_pack(a := 1, b := 2)",
+            write={"presto": "CAST(ROW(1, 2) AS ROW(a INTEGER, b INTEGER))"},
+        )
+
+        self.validate_all("struct_pack(a := 1, b := x)", write={"presto": UnsupportedError})
```
Comment (georgesittas) [inline tests/dialects/test_duckdb.py:66]: let's also test the roundtrip to duckdb

File: tests/dialects/test_hive.py [hunk 4]
```diff
@@ -611,12 +611,6 @@ def test_hive(self):
                 "spark": "GET_JSON_OBJECT(x, '$.name')",
             },
         )
-        self.validate_all(
-            "STRUCT(a = b, c = d)",
-            read={
-                "snowflake": "OBJECT_CONSTRUCT(a, b, c, d)",
-            },
-        )
```
Comment (georgesittas) [inline tests/dialects/test_hive.py:619]: why was this removed?

File: tests/dialects/test_snowflake.py [hunk 5]
```diff
@@ -372,13 +372,13 @@ def test_snowflake(self):
             write={"snowflake": "SELECT * FROM (VALUES (0)) AS foo(bar)"},
         )
         self.validate_all(
-            "OBJECT_CONSTRUCT(a, b, c, d)",
+            "OBJECT_CONSTRUCT('a', b, 'c', d)",
```
Comment (georgesittas) [inline tests/dialects/test_snowflake.py:375]: fyi `OBJECT_CONSTRUCT(a, b)` is also valid snowflake:

```
-- yields { "foo": 1 }
WITH t(a) AS (SELECT 'foo')
SELECT OBJECT_CONSTRUCT(a, 1) from t
```

**Comment by georgesittas** [comment]
Nice!

File: tests/dialects/test_hive.py (hunk 4 above)
Comment (tobymao) [reply tests/dialects/test_hive.py:619]: didn't find any docs on hive struct

File: tests/dialects/test_bigquery.py (hunk 2 above)
Comment (tobymao) [reply tests/dialects/test_bigquery.py:617]: the issue is if you do it it's an unnamed struct so i changed it

File: tests/dialects/test_hive.py (hunk 4 above)
Comment (tobymao) [reply tests/dialects/test_hive.py:619]: added unnamed structs for hive, it doesn't look like it uses key value

File: sqlglot/parser.py [hunk 6]
```diff
@@ -4132,6 +4137,26 @@ def _parse_function_call(
         self._match_r_paren(this)
         return self._parse_window(this)
 
+    def _kv_to_prop_eq(self, expressions: t.List[exp.Expression]) -> t.List[exp.Expression]:
+        transformed = []
+
+        for e in expressions:
+            if isinstance(e, self.KEY_VALUE_DEFINITIONS):
+                if isinstance(e, exp.Alias):
+                    e = self.expression(exp.PropertyEQ, this=e.args.get("alias"), expression=e.this)
```
Comment (georgesittas) [inline sqlglot/parser.py:4146]:
```suggestion
                    e = self.expression(exp.PropertyEQ, this=e.args["alias"], expression=e.this)
```

File: sqlglot/parser.py [hunk 7]
```diff
@@ -4132,6 +4137,26 @@ def _parse_function_call(
...
+
+                if not isinstance(e, exp.PropertyEQ):
+                    e = self.expression(
+                        exp.PropertyEQ, this=exp.to_identifier(e.name), expression=e.expression
+                    )
+
+                if isinstance(e.this, exp.Column):
+                    e.this.replace(e.this.this)
```
Comment (georgesittas) [inline sqlglot/parser.py:4154]: Just double-checking, but we replace here so that we only keep the identifier right?

File: sqlglot/dialects/hive.py [hunk 8]
```diff
@@ -631,3 +631,15 @@ def datatype_sql(self, expression: exp.DataType) -> str:
         def version_sql(self, expression: exp.Version) -> str:
             sql = super().version_sql(expression)
             return sql.replace("FOR ", "", 1)
+
+        def struct_sql(self, expression: exp.Struct) -> str:
+            values = []
+
+            for i, e in enumerate(expression.expressions):
+                if isinstance(e, exp.PropertyEQ):
+                    self.unsupported("Hive does not support named structs.")
```
Comment (georgesittas) [inline sqlglot/dialects/hive.py:640]: I know this is gonna be annoying, but FWIW Spark v2 supports `STRUCT(1 AS a)` (so maybe this is also supported by Hive?) and also I found an [SO thread](https://stackoverflow.com/a/32386011) that shows a different function `select named_struct("a", 1)`, supposedly supported by Hive as well AFAICT. Works in both Spark v2 and v3.

File: sqlglot/dialects/spark2.py [hunk 9]
```diff
@@ -293,5 +283,10 @@ def columndef_sql(self, expression: exp.ColumnDef, sep: str = " ") -> str:
                 ),
             )
 
+        def struct_sql(self, expression: exp.Struct) -> str:
+            from sqlglot.generator import Generator
+
+            return Generator.struct_sql(self, expression)
```
Comment (georgesittas) [inline sqlglot/dialects/spark2.py:289]: Can also add this in `TRANSFORMS` if we wanna avoid defining this method? It should take precedence.

**Review by georgesittas — COMMENTED** [review]
I think we may wanna add another test for snowflake to check that
```
OBJECT_CONSTRUCT(a, b, c, d)
```
can be roundtripped.

File: sqlglot/parser.py (hunk 7 above)
Comment (tobymao) [reply sqlglot/parser.py:4154]: yea, we don't want a column, it will get qualified etc, it's an identifier

File: sqlglot/dialects/hive.py (hunk 8 above)
Comment (tobymao) [reply sqlglot/dialects/hive.py:640]: no, that's a plugin

**Review by georgesittas — APPROVED** [review]
